compiler/rustc_data_structures/src/base_n.rs

```diff
@@ -42,3 +43,52 @@ pub fn encode(n: u128, base: usize) -> String {
     push_str(n, base, &mut s);
     s
 }
+
+// This trait just lets us reserve the exact right amount of space when doing fixed-length
+// case-insensitve encoding. Add any impls you need.
+pub trait Base36Encodable: Copy + Into<u128> {
+    fn encoded_len() -> usize;
+}
+
+impl Base36Encodable for u128 {
+    fn encoded_len() -> usize {
+        25
+    }
+}
+
+impl Base36Encodable for u64 {
+    fn encoded_len() -> usize {
+        13
+    }
+}
+
+impl Base36Encodable for u32 {
+    fn encoded_len() -> usize {
+        7
+    }
+}
+
+pub fn push_case_insensitive<N: Base36Encodable>(n: N, output: &mut String) {
```

Comment:
This need some comment about what the difference between this fn and push_str, except hardcoded base.